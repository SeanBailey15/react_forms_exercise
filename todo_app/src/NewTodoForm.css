.NewTodoForm {
  display: flex;
  justify-content: center;
  width: 65vw;
}

.NewTodoForm-form {
  display: flex;
  flex-direction: column;
  width: 50%;
}

.NewTodoForm-label {
  text-align: center;
  margin-bottom: 5px;
}

.NewTodoForm-input {
  text-align: center;
  margin-bottom: 5px;
  width: 100%;
}

.NewTodoForm-err {
  text-align: center;
  color: red;
  font-weight: bolder;
  margin-bottom: 5px;
}

.NewTodoForm-btn {
  border: 3px solid black;
  border-radius: 10px;
  font-weight: bolder;
  margin-top: 10px;
  background-color: palegreen;
}
