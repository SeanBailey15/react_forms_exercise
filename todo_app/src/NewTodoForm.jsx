import React from "react";
import { useFormik } from "formik";
import './NewTodoForm.css'

const validate = values => {
    const errors = {};
    if(!values.content){
        errors.content = 'Required';
    } else if(values.content.length > 30){
        errors.content = 'Todo Must Be 30 Characters Or Less'
    }
    return errors;
}

const NewTodoForm = ({addTodo}) => {
    const formik = useFormik({
        initialValues: {
            content: ''
        },
        validate,
        onSubmit: values => {
            addTodo({...values})
            formik.resetForm()
        }
    })

    return (
        <div className="NewTodoForm">
            <form className="NewTodoForm-form" onSubmit={formik.handleSubmit}>
                <label className="NewTodoForm-label" htmlFor="content" >Todo</label>
                <input
                    className="NewTodoForm-input" 
                    id="content"
                    name="content"
                    type="text"
                    placeholder="What to do, what to do?"
                    value={formik.values.content}
                    onChange={formik.handleChange}
                />
                {formik.errors.content && formik.touched.content && (
                    <div className="NewTodoForm-err">{formik.errors.content}</div>
                )}
                <button className="NewTodoForm-btn" type="submit" onClick={formik.handleSubmit}>Add A Task</button>
            </form>
        </div>
    )
}

export default NewTodoForm