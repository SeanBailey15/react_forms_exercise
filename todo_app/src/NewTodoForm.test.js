import React from "react";
import { render, fireEvent, screen, act } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";

it("renders without crashing", () => {
  act(() => {
    render(<NewTodoForm />);
  });
});

it("matches snapshot", () => {
  const { asFragment } = render(<NewTodoForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("tests for empty input", async () => {
  act(() => {
    render(<NewTodoForm />);
  });

  const content = screen.getByLabelText("Todo");
  const addButton = screen.getByText("Add A Task");

  // Initially, the error message should not be present
  expect(screen.queryByText("Required")).not.toBeInTheDocument();

  // Try to submit an empty input
  fireEvent.click(addButton);

  // Wait for the error message to appear and check if it is present
  const errorMessage = await screen.findByText("Required");
  expect(errorMessage).toBeInTheDocument();
});

//************************************************************** */
// Repeat above steps for other expected validation errors

it("tests for input greater than 30 characters", async () => {
  act(() => {
    render(<NewTodoForm />);
  });

  const content = screen.getByLabelText("Todo");
  const addButton = screen.getByText("Add A Task");

  expect(
    screen.queryByText("Todo Must Be 30 Characters Or Less")
  ).not.toBeInTheDocument();

  fireEvent.change(content, {
    target: { value: "I am a very long todo that will not pass validation" },
  });
  fireEvent.click(addButton);

  const errorMessage = await screen.findByText(
    "Todo Must Be 30 Characters Or Less"
  );
  expect(errorMessage).toBeInTheDocument();
});
