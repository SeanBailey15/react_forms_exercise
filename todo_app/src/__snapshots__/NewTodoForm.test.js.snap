// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="NewTodoForm"
  >
    <form
      class="NewTodoForm-form"
    >
      <label
        class="NewTodoForm-label"
        for="content"
      >
        Todo
      </label>
      <input
        class="NewTodoForm-input"
        id="content"
        name="content"
        placeholder="What to do, what to do?"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-btn"
        type="submit"
      >
        Add A Task
      </button>
    </form>
  </div>
</DocumentFragment>
`;
